pip/pip-359.md

```diff
@@ -0,0 +1,213 @@
+                .topic("persistent://public/default/my-topic")
+                .subscriptionName("my-subscription2")
+                .messageListener((consumer, message) -> {
+                    // process message
+                    consumer.acknowledgeAsync(message);
+                })
+                .subscribe();
+    }
+}
+```
+
+# Motivation
+
+As [Background knowledge](#background-knowledge) mentioned, when using asynchronous consumer, 
+by registering a MessageListener interface, there is a problem of different consumer groups 
+affecting each other, leading to unnecessary consumption delays.
+Therefore, for this scenario, this PIP prepare to support specific a message
+listener executor of consumer latitudes to solve such problem.
+# Goals
+1. Improve consumer message listener isolation, solve the consumption delay problem caused by 
+mutual influence of different consumers from the same PulsarClient instance.
+
+## In Scope
+If this PIP is accepted, it will help Pulsar solve the problem of different consumers 
+from same `PulsarClient` affecting each other in the asynchronous consumption mode(`MessageListener`).
+
+## Out of Scope
+This PIP will not build the plugin library mentioned in [PR](https://github.com/apache/pulsar/pull/22902#issuecomment-2169962642),
+we will open a new PIP in the future to do this
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. Add an interface `MessageListenerExecutor`, responsible for executing message listener callback tasks.
+Users can customize the implementation to determine in which thread the message listener task is executed.
+For example, in the situation described in [Motivation](#motivation) part, users can implement the 
+interface with an independent underlying thread pool to ensure that the message listener task of each 
+consumer is executed in a separate thread. The caller would be responsible for the life cycle of the 
+Executor, and it would be used only for this specific consumer.
+    ```java
+    public interface MessageListenerExecutor {
+
+    /**
+     * select a thread by message(if necessary, for example, 
+     * Key_Shared SubscriptionType, maybe need select thread 
+     * by message order key to ensure order) to execute the runnable!
+     *
+     * @param message  the message
+     * @param runnable the runnable to execute
+     */
+    void execute(Message<?> message, Runnable runnable);
+    }
+    ```
+2. Add an optional config `messageListenerExecutor` in `ConsumerBuilder`, then 
+users can pass their implementations.
+    ```java
+    ConsumerBuilder<T> messageListenerExecutor(MessageListenerExecutor messageListenerExecutor);
+    ```
+
+### Why need an interface like `MessageListenerExecutor`
+Some people may wonder why not just use `java.util.concurrent.ExecutorService`,
+but define an interface like `MessageListenerExecutor`. 
+
+The reason is that:
+
+For sequential consumption scenarios, we need to ensure that messages with the same 
```

Comment:
Why is it important that same partition or same key is processed by the same thread? Why just processing them in order will not suffice? 
The only thing you need is for the same partition to not send it to execute before the previous one finished, right?

Reply:
> Why is it important that same partition or same key is processed by the same thread? Why just processing them in order will not suffice? The only thing you need is for the same partition to not send it to execute before the previous one finished, right?

If messages in the same partition cannot be processed by the same thread, we will not be able to guarantee the order of consumption unless we use locks. 

Assume the following scenario, we need to ensure that the three messages m1, m2, and m3 from order-topic-partition-1 are consumed in order:
Usually, three messages correspond to three messageListener callback tasks. If we submit these three tasks to a thread pool containing many threads, then these three tasks can be executed at the same time, thus destroying the order. 
However, if we submit these three callback tasks to the same thread pool in order, and this thread pool only contain one thread, and then the three callback tasks will be processed one by one, thus ensuring the order.

Reply:
> Why is it important that same partition or same key is processed by the same thread? Why just processing them in order will not suffice? The only thing you need is for the same partition to not send it to execute before the previous one finished, right?

I also added two figures to help illustrate this problem, which you can check in the latest PIP-359.